pkg/autoupdate/tuf/library_manager.go

```diff
@@ -0,0 +1,359 @@
+	logger          log.Logger
+}
+
+func newUpdateLibraryManager(metadataClient *client.Client, mirrorUrl string, mirrorClient *http.Client, baseDir string, operatingSystem string, osquerier localserver.Querier, logger log.Logger) (*updateLibraryManager, error) {
+	ulm := updateLibraryManager{
+		metadataClient:  metadataClient,
+		mirrorUrl:       mirrorUrl,
+		mirrorClient:    mirrorClient,
+		baseDir:         baseDir,
+		operatingSystem: operatingSystem,
+		osquerier:       osquerier,
+		logger:          log.With(logger, "component", "tuf_autoupdater_library_manager"),
+	}
+
+	// Ensure the updates directory exists
+	if err := os.MkdirAll(baseDir, 0755); err != nil {
+		return nil, fmt.Errorf("could not make base directory for updates library: %w", err)
+	}
+
+	// Ensure our staged updates and updates directories exist
+	for _, binary := range binaries {
+		// Create the directory for staging updates
+		if err := os.MkdirAll(ulm.stagedUpdatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make staged updates directory for %s: %w", binary, err)
+		}
+
+		// Create the update library
+		if err := os.MkdirAll(ulm.updatesDirectory(binary), 0755); err != nil {
+			return nil, fmt.Errorf("could not make updates directory for %s: %w", binary, err)
+		}
+	}
+
+	return &ulm, nil
+}
+
+// updatesDirectory returns the update library location for the given binary.
+func (ulm *updateLibraryManager) updatesDirectory(binary autoupdatableBinary) string {
+	return filepath.Join(ulm.baseDir, string(binary))
+}
+
+// stagedUpdatesDirectory returns the location for staged updates -- i.e. updates
+// that have been downloaded, but not yet verified.
+func (ulm *updateLibraryManager) stagedUpdatesDirectory(binary autoupdatableBinary) string {
+	return filepath.Join(ulm.baseDir, fmt.Sprintf("%s-staged", binary))
+}
+
+// addToLibrary adds the given target file to the library for the given binary,
+// downloading and verifying it if it's not already there. After any addition
+// to the library, it cleans up older versions that are no longer needed.
+func (ulm *updateLibraryManager) AddToLibrary(binary autoupdatableBinary, targetFilename string) error {
+	// Check to see if the current running version is the version we were requested to add;
+	// return early if it is, but don't error out if we can't determine the current version.
+	currentVersion, err := ulm.currentRunningVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get current running version", "binary", binary, "err", err)
+	} else {
+		if currentVersion.Original() == ulm.versionFromTarget(binary, targetFilename) {
+			// We don't need to download the current running version because it already exists,
+			// either in this updates library or in the original install location.
+			return nil
+		}
+	}
+
+	if ulm.alreadyAdded(binary, targetFilename) {
+		return nil
+	}
+
+	defer ulm.tidyLibrary(binary, currentVersion)
```

Comment:
TODO tidying should happen on startup and not post-update to avoid accidentally removing an in-use intermediate binary

Reply:
Would it make sense to split the "tidying" responsibility out? This is a large-ish file, maybe the "tidier" could be injected in and run immediately, or guaranteed to run before the manager does it's thing.

Reply:
Maybe a `LibraryCustodian` 🤔 

Yeah that's reasonable, this is getting a little bit unwieldy. Let me see what makes sense for splitting it up!

Reply:
Both adding to the library and tidying the library require a) being able to query the current running version and b) knowing where to find the updates directories and staged update directories per binary. I couldn't find a way to split this up (given the shared knowledge and responsibilities) where the abstraction felt meaningful.

I think in the future we'll have the tidying routine run very early on, right after version selection for launcher and osqueryd, and then it might feel a little easier to refactor?